### ISSUE TITLE
server: decommissioning starts draining and then cancels it - I think we're not really draining when decomissioning

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#25586]
Here's a nice one. I have a change that makes `TestHeartbeatCallbackForDecommissioning` flaky. I have no idea how the change is affecting it exactly, perhaps other than weirdly affecting some timings.
But I've dug into what happens and the code just seems broken to me:
A server's liveness heartbeat loop detects the decommissioning bit on the liveness record and triggers draining. Except that it does draining as an async `Worker` and then races to cancel the context used by the worker (lolz). Canceling that context can and does stop the draining.
Here's some code:
This is the worker doing the draining: https://github.com/cockroachdb/cockroach/blob/82b4293e3aff8c50952fddacbb22b57d34d1044c/pkg/server/server.go?utf8=%E2%9C%93#L1500

Notice that it's called from a closure passed to `nodeLiveness.StartHeartbeat()`. Now, the callback is called very indirectly from the heartbeat loop, but here's the context that's being used for it:
https://github.com/cockroachdb/cockroach/blob/82b4293e3aff8c50952fddacbb22b57d34d1044c/pkg/storage/node_liveness.go#L451

As I was saying, this context cancelation races with callback. The test becomes flaky because sometimes we cancel so soon that we don't even get to mark the liveness record as "draining". Even when the test is not flaky, I assume that we don't actually do the "leases" part of server draining - i.e. the lease transfers. I postulate that decommissioning is thus quite broken - but I don't have proof.
If true, some sort of acceptance test might be in order.

Probably an immediate fix is to remove the parent's cancelation from the ctx that's passed to the draining worker. But I don't know if that's the right fix.

**Comment by tschottdorf** [comment]
The first link you posted shows that we pass `context.Background()` into `SetDraining`, preceded by a comment on why we don't use the passed-in context. Your comment reads like you missed that, is that correct or am I the one missing things here?

**Comment by andreimatei** [comment]
I'm an imbecile, I'm sorry.
I've removed that `context.Background()` from my PR and forgot that I did. I had removed it because I thought that its point was to work around the fact that the `TxnCoordSender` is capturing the context of the first txn in the operation and uses it for the heartbeat loop - which is one of the things that the PR in question is trying to fix. I guess I've discovered why the `context.Background()` was really there.
